### PR TITLE
Framegraph: add renderpass image accesses

### PR DESCRIPTION
**Pull request by hevrard**
List the images accessed by a renderpass.

Refactor a bit the types to extract imageInfo as a standalone type
from attachmentInfo.

Use a lookup table to find which images match a given memory access.

Bug: b/172218645
Test: manual, on several traces

### CODE REVIEW

**Comment by hevrard** [comment]
This PR leads to the following framegraph on SW "deferred" demo:

framegraph

**Review by johnplate — APPROVED** [review]
LGTM, but again wait for another approval.

File: gapis/api/vulkan/framegraph.go
Comment (yalcinmelihyasin) [inline gapis/api/vulkan/framegraph.go:102]: Very nit: You probably don't need the variable.

File: gapis/api/vulkan/framegraph.go
Comment (yalcinmelihyasin) [inline gapis/api/vulkan/framegraph.go:215]: Very nit: I think that writing "memoization table" can save a lot of word.

File: gapis/api/vulkan/framegraph.go
Comment (yalcinmelihyasin) [inline gapis/api/vulkan/framegraph.go:227]: This looks like a very deeply accessed information that can cause consecutive cache misses(e.g accessing the elements of two layered map). If there is no alternative, then it's fine but otherwise, for the future, this structure can be reconsidered.

Comment (yalcinmelihyasin) [reply gapis/api/vulkan/framegraph.go:227]: This of course depends how it's being read.

File: gapis/api/vulkan/framegraph.go
Comment (yalcinmelihyasin) [inline gapis/api/vulkan/framegraph.go:248]: If we are looping over a map to find if a range is included in a memory range, maybe we should store them as array under a pool instead of a map. Because if we are not using the direct access of the map(and I believe these are hash-tables) then we should not pay the cost for non-consecutive memory access. In a large enough data set, it should be much faster to loop over an array than a map.

Unless, there will be another access to the memory range that will be added in the future that queries the ranges with directly accessing to the map elements.

File: gapis/api/vulkan/framegraph.go
Comment (yalcinmelihyasin) [inline gapis/api/vulkan/framegraph.go:238]: We haven't check if the range is already existing or any existing range includes the current range. If I understand correctly, what happens here behind the scene is;

- check if memRange exist for the map stored imageLookup[pool] 
- if exists add &imgObj to the slice stored in imageLookup[pool][memRange]
- if not, then append imgObj to a nil slice(which creates it) then assign that slice to imageLookup[pool][memRange] which again allocates some stuff on the map.

If this is the intended behaviour then it's ok. Otherwise, please reconsider.

File: gapis/api/vulkan/framegraph.go
Comment (yalcinmelihyasin) [inline gapis/api/vulkan/framegraph.go:250]: can one memRange included by different memRanges in the map?

File: gapis/api/vulkan/framegraph.go
Comment (yalcinmelihyasin) [inline gapis/api/vulkan/framegraph.go:306]: You may want to cache this unless you are modifying it in the loop. I am not sure, double check this, but go may try to call dependencyGraph.GetNodeAccesses(nodeID) in every cycle in the loop.

**Review by yalcinmelihyasin — COMMENTED** [review]
I don't have a very deep understanding of how memory ranges works that is used in this PR. Therefore some comments that I made can be not perfectly valid.

There are a few things that I wanted to point out and if those questions are known or not an issue, then it's OK to merge this PR with LGTM from another person.

PS: Nit & Very Nit are optional.

File: gapis/api/vulkan/framegraph.go
Comment (hevrard) [reply gapis/api/vulkan/framegraph.go:102]: Turns out you do, because Go won't let you do `&(imgView.Image())` (hence this new variable here).

File: gapis/api/vulkan/framegraph.go
Comment (hevrard) [reply gapis/api/vulkan/framegraph.go:215]: Memoization is more about lazily computing results, storing them, and then lookup them up next time they are queried. I think lookup table is more accurate here since we populate the entire table upfront, and then we always use it for lookups. I also feel that it's semantics details at this point :)

File: gapis/api/vulkan/framegraph.go
Comment (hevrard) [reply gapis/api/vulkan/framegraph.go:306]: Go will evaluate it only once, the Go spec says (https://golang.org/ref/spec#For_statements , section For statements with range clause):

> The range expression x is evaluated once before beginning the loop, with one exception: if at most one iteration variable is present and len(x) is constant, the range expression is not evaluated.

File: gapis/api/vulkan/framegraph.go
Comment (hevrard) [reply gapis/api/vulkan/framegraph.go:250]: Here in theory there could be several imgRange that could include memRange, yes. Maybe not in practice, but it doesn't hurt to check them all, since in practice there are almost always only one range associated to a poolID (so the loop on `helpers.imageLookup[pool]` will have only one iteration).

File: gapis/api/vulkan/framegraph.go
Comment (hevrard) [reply gapis/api/vulkan/framegraph.go:227]: In practice, given how memory accesses are recorded at capture time, there is almost always only one memory range associated to a memory pool. In the lookup, we know the poolID so we directly access to the second-level map, then we iterate over it and it's not a big deal since it will almost always have only one element.

File: gapis/api/vulkan/framegraph.go
Comment (hevrard) [reply gapis/api/vulkan/framegraph.go:248]: I use a second-level map indexed by ranges in order to avoid storing duplicate ranges. Again, in practice for each poolID we almost always have only a single range anyway.

File: gapis/api/vulkan/framegraph.go
Comment (hevrard) [reply gapis/api/vulkan/framegraph.go:238]: You got the idea, we append to helpers.imageLookup[pool][memRange], which may be nil.

This seems to be the classic way to add items to a slice member of a map, quoting https://blog.golang.org/maps:

> Another instance of helpful zero values is a map of slices. Appending to a nil slice just allocates a new slice, so it's a one-liner to append a value to a map of slices; there's no need to check if the key exists.

**Pull request merged**
